substrate/frame/nomination-pools/src/lib.rs

```diff
@@ -2823,6 +2849,58 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Apply a pending slash on a member.
+		///
+		/// This call can be dispatched permissionlessly (i.e. by any account). If the member has
+		/// slash to be applied, caller may be rewarded with the part of the slash.
+		#[pallet::call_index(23)]
+		#[pallet::weight(T::WeightInfo::apply_slash())]
+		pub fn apply_slash(
+			origin: OriginFor<T>,
+			member_account: AccountIdLookupOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let member_account = T::Lookup::lookup(member_account)?;
+			Self::do_apply_slash(&member_account, Some(who))?;
+
+			// If successful, refund the fees.
+			Ok(Pays::No.into())
+		}
+
+		/// Migrates delegated funds from the pool account to the `member_account`.
+		///
+		/// This is a permission-less call and refunds any fee if claim is successful.
+		///
+		/// If the pool has migrated to delegation based staking, the staked tokens of pool members
+		/// can be moved and held in their own account. See [`adapter::DelegateStake`]
+		#[pallet::call_index(24)]
+		#[pallet::weight(T::WeightInfo::migrate_delegation())]
+		pub fn migrate_delegation(
+			origin: OriginFor<T>,
+			member_account: AccountIdLookupOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let _caller = ensure_signed(origin)?;
+
+			let member_account = T::Lookup::lookup(member_account)?;
+			let member =
+				PoolMembers::<T>::get(&member_account).ok_or(Error::<T>::PoolMemberNotFound)?;
+			let pool_contribution = member.total_balance();
+			ensure!(pool_contribution >= MinJoinBond::<T>::get(), Error::<T>::MinimumBondNotMet);
+
+			let delegation = T::StakeAdapter::member_delegation_balance(&member_account);
+			ensure!(pool_contribution > delegation, Error::<T>::NoDelegationToClaim);
+
+			let diff = pool_contribution.defensive_saturating_sub(delegation);
+			T::StakeAdapter::migrate_delegation(
+				&Pallet::<T>::generate_bonded_account(member.pool_id),
```

Comment:
these APIs are tricky in that you might mistakenly swap params.. but okay.

Reply:
Yeah. I think we talked about wrapping them in a new type. It should be a good improvement, will make the change.